File: ios/Kurir/modules/contacts/ContactsModule.swift
//
//  ContactsModule.swift
//  Kurir
//
//  Created by Loren on 19/9/19.
//  Copyright © 2019 Crazybean Studio. All rights reserved.
//

import Mobilex

class ContactsModule: Module {
    override func inject() {
        // Wrapper
        factory(ContactsWrapper.self) { r in
            ContactsWrapper(userData: r.resolve(UserData.self),
                            usersRepository: r.resolve(UsersRepository.self),
                            contactsRepository: r.resolve(ContactsRepository.self))
        }
        
        // Actor
        factory(ContactsActor.self) { (r, scene: ContactsScene) -> ContactsActor in
            ContactsActor(scene: scene, wrapper: r.resolve(ContactsWrapper.self)!)
        }
    }
}

File: android/src/main/java/au/com/crazybean/mobilex/kurir/modules/contacts/ContactsModule.kt
package au.com.crazybean.mobilex.kurir.modules.contacts

import au.com.crazybean.mobilex.kurir.extension.qualifier
import au.com.crazybean.mobilex.kurir.modules.base.ViewModel
import org.koin.androidx.viewmodel.dsl.viewModel
import org.koin.dsl.module

val contactsModule = module {
    val qualifier = qualifier<ContactsWrapper>()

    // Wrapper via ViewModel
    viewModel(qualifier) {
        ViewModel(ContactsWrapper(get(), get(), get()))
    }

    // Actor
    factory { (scene: ContactsScene?) ->
        val viewModel = get<ViewModel<ContactsWrapper>>(qualifier)
        ContactsActor(scene, viewModel.wrapper)
    }
}